src/ProgressOnderwijsUtils/Data/MetaObjectDataReader.cs

```diff
@@ -115,23 +115,53 @@ public ColumnInfo([NotNull] IReadonlyMetaProperty<T> mp)
+                    var columnBoxedAsObject = Expression.Convert(columnValueAsNonNullable, typeof(object));
+                    Expression columnBoxedAsColumnType;
+                    if (isNonNullable) {
+                        columnBoxedAsColumnType = columnBoxedAsObject;
+                        WhenNullable_IsColumnDBNull = null;
+                    } else {
+                        var propertyIsNotNull = IsExpressionNonNull(propertyValue);
+                        columnBoxedAsColumnType = Expression.Condition(propertyIsNotNull, columnBoxedAsObject, Expression.Constant(DBNull.Value, typeof(object)));
+                        WhenNullable_IsColumnDBNull = Expression.Lambda<Func<T, bool>>(Expression.Not(propertyIsNotNull), metaObjectParameter).Compile();
+                    }
+
+                    TypedNonNullableGetter = Expression.Lambda(typeof(Func<,>).MakeGenericType(typeof(T), propertyConverter.DbType), columnValueAsNonNullable, metaObjectParameter).Compile();
+                    GetUntypedColumnValue = Expression.Lambda<Func<T, object>>(columnBoxedAsColumnType, metaObjectParameter).Compile();
                 } else {
-                    var propertyIsDefault = Expression.Equal(Expression.Default(propertyType), propertyValue);
-                    columnBoxedAsColumnType = Expression.Coalesce(propertyConvertedToUnderlyingType, Expression.Constant(DBNull.Value, typeof(object)));
-                    WhenNullable_IsColumnDBNull = Expression.Lambda<Func<T, bool>>(propertyIsDefault, metaObjectParameter).Compile();
+                    ColumnType = nonNullableUnderlyingType;
+                    var propertyValueAsNoNullable = Expression.Convert(propertyValue, ColumnType);
+                    var columnValueAsNonNullable = Expression.Convert(propertyValue, propertyType.GetUnderlyingType());
+                    var columnBoxedAsObject = Expression.Convert(columnValueAsNonNullable, typeof(object));
+                    Expression columnBoxedAsColumnType;
+                    if (isNonNullable) {
+                        columnBoxedAsColumnType = columnBoxedAsObject;
+                        WhenNullable_IsColumnDBNull = null;
+                    } else {
+                        var propertyIsNotNull = IsExpressionNonNull(propertyValue);
+                        columnBoxedAsColumnType = Expression.Coalesce(columnValueAsNonNullable, Expression.Constant(DBNull.Value, typeof(object)));
+                        WhenNullable_IsColumnDBNull = Expression.Lambda<Func<T, bool>>(Expression.Not(propertyIsNotNull), metaObjectParameter).Compile();
+                    }
+                    TypedNonNullableGetter = Expression.Lambda(typeof(Func<,>).MakeGenericType(typeof(T), ColumnType), propertyValueAsNoNullable, metaObjectParameter).Compile();
+                    GetUntypedColumnValue = Expression.Lambda<Func<T, object>>(columnBoxedAsColumnType, metaObjectParameter).Compile();

```

Comment:
Op den duur willen we denk ik wel proberen de verschillen tussen deze branches weg te werken, en tevens ook de valueConverter voor readers te gebruiken, maar vooralsnog: prima!